pkg/schemadsl/compiler/compiler.go

```diff
@@ -19,37 +21,73 @@ type InputSchema struct {
 	SchemaString string
 }
 
+// SchemaDefinition represents an object or caveat definition in a schema.
+type SchemaDefinition interface {
+	proto.Message
+
+	GetName() string
+}
+
+// CompiledSchema is the result of compiling a schema when there are no errors.
+type CompiledSchema struct {
+	// ObjectDefinitions holds the object definitions in the schema.
+	ObjectDefinitions []*core.NamespaceDefinition
+
+	// CaveatDefinitions holds the caveat definitions in the schema.
+	CaveatDefinitions []*core.CaveatDefinition
+
+	// OrderedDefinitions holds the object and caveat definitions in the schema, in the
+	// order in which they were found.
+	OrderedDefinitions []SchemaDefinition
+}
+
 // Compile compilers the input schema(s) into a set of namespace definition protos.
-func Compile(schemas []InputSchema, objectTypePrefix *string) ([]*core.NamespaceDefinition, error) {
+func Compile(schemas []InputSchema, objectTypePrefix *string) (*CompiledSchema, error) {
 	mapper := newPositionMapper(schemas)
 
 	// Parse and translate the various schemas.
-	definitions := []*core.NamespaceDefinition{}
+	orderedDefinitions := []SchemaDefinition{}
+
 	for _, schema := range schemas {
 		root := parser.Parse(createAstNode, schema.Source, schema.SchemaString).(*dslNode)
 		errs := root.FindAll(dslshape.NodeTypeError)
 		if len(errs) > 0 {
 			err := errorNodeToError(errs[0], mapper)
-			return []*core.NamespaceDefinition{}, err
+			return nil, err
 		}
 
 		translatedDefs, err := translate(translationContext{
 			objectTypePrefix: objectTypePrefix,
 			mapper:           mapper,
+			schemaString:     schema.SchemaString,
 		}, root)
 		if err != nil {
 			var errorWithNode errorWithNode
 			if errors.As(err, &errorWithNode) {
 				err = toContextError(errorWithNode.error.Error(), errorWithNode.errorSourceCode, errorWithNode.node, mapper)
 			}
 
-			return []*core.NamespaceDefinition{}, err
+			return nil, err
 		}
 
-		definitions = append(definitions, translatedDefs...)
+		orderedDefinitions = append(orderedDefinitions, translatedDefs...)
 	}
 
-	return definitions, nil
+	return &CompiledSchema{
+		CaveatDefinitions:  filterTo[*core.CaveatDefinition](orderedDefinitions),
+		ObjectDefinitions:  filterTo[*core.NamespaceDefinition](orderedDefinitions),
+		OrderedDefinitions: orderedDefinitions,
+	}, nil
+}
+
+func filterTo[T SchemaDefinition](orderedDefinitions []SchemaDefinition) []T {
```

Comment:
If you mean why not just return slices of namespaces and caveats, its because we need to maintain the ordered definitions as well for generation, and I figured that returning three slices was getting to be a bit too much.

Let me know if you prefer the three slices to this approach.

Reply:
As discussed, changed it to return the concrete type instead